src/plugins/intel_cpu/src/graph_optimizer.cpp

```diff
@@ -2398,10 +2401,165 @@ void GraphOptimizer::FusePerformedAsScaleShiftAndFakeQuantize(Graph &graph) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void GraphOptimizer::mergeTransposeReshapeReorder(Graph& graph,
+                                                  const NodePtr& transposeNode,
+                                                  const NodePtr& reshapeNode,
+                                                  const NodePtr& reorderNode,
+                                                  const bool reverseOrder) {
+    const auto& parentNode = reverseOrder ? reorderNode : transposeNode;
+    const auto& childNode = reverseOrder ? transposeNode : reorderNode;
+    auto nodeBeforeSequence = parentNode->getParentEdgesAtPort(0)[0]->getParent();
+    auto nodeAfterSequence = childNode->getChildEdgeAt(0)->getChild();
+
+    auto removeInputEdge = [&](const NodePtr& node, const size_t idx) {
+        auto remEdge = node->getParentEdgesAtPort(idx)[0];
+        auto parentNode = remEdge->getParent();
+        remEdge->drop();
+        auto& edges = graph.GetEdges();
+        for (auto it = edges.begin(); it != edges.end(); it++) {
+            if ((*it) == remEdge) {
+                edges.erase(it);
+                if (parentNode->getChildEdges().empty())
+                    parentNode->remove();
+                break;
+            }
+        }
+    };
+
+    removeInputEdge(transposeNode, 1);
+    if (reshapeNode)
+        removeInputEdge(reshapeNode, 1);
+
+    // to prevent inPlace conflict we must check that the memory reference is unidirectional or
+    // inPlace memory is not used
+    const auto parentInPlace = parentNode->getParentEdgeAt(0)->inPlace(Edge::LOOK_UP);
+    const auto& childEdges = childNode->getChildEdgesAtPort(0);
+    const auto childInPlace = std::any_of(childEdges.begin(), childEdges.end(), [](const EdgePtr& edge) {
+        return edge->inPlace(Edge::LOOK_DOWN);
+    });
+    bool isOptimized = !(parentInPlace && childInPlace);
+
+    graph.DropNode(transposeNode);
+    graph.DropNode(reorderNode);
+    if (reshapeNode)
+        graph.DropNode(reshapeNode);
+
+    auto inDesc = parentNode->getSelectedPrimitiveDescriptor()->getConfig().inConfs[0].getMemDesc();
+    auto outDesc = childNode->getSelectedPrimitiveDescriptor()->getConfig().outConfs[0].getMemDesc();
+
+    auto inPrec = inDesc->getPrecision();
+    auto outPrec = outDesc->getPrecision();
+
+    auto reorderInDesc = inDesc;
+    auto reorderOutDesc = outDesc->cloneWithNewPrecision(inPrec);
+
+    EdgePtr edge;
+    for (auto& childEdge : nodeBeforeSequence->getChildEdges()) {
+        if (childEdge.lock()->getChild() == nodeAfterSequence) {
+            edge = childEdge.lock();
+            break;
+        }
+    }
+    if (!edge) {
+        IE_THROW() << "Parent node '" << parentNode->getName() << "' has invalid edges.";
+    }
+
+    // transposeNode support blocked input & non-blocked output, in the case, the Reorder after Transpose cannot be optimized
+    std::vector<int> srcPerm;
+    if (!reverseOrder) {
```

Comment:
I think this permute step check is asked for both case of Transpose+Reorder and Reorder+Transpose. We need corresponding unit tests to trigger the conditions for reverseOrder==True case, and address it here.

Reply:
As described in the comment, _**"transposeNode support blocked input & non-blocked output, in the case, the Reorder after Transpose cannot be optimized"**_. But we cannot have the opposite situation (non-blocked input but blocked output) (I saw [Transpose::initSupportedPrimitiveDescriptors](https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/src/nodes/transpose.cpp#L98)). Based on this I concluded that in case of reverse order we don't need this check. Am I wrong?

Reply:
Considering two situations -
1.  (some layout A) -> Reorder -> (blocked layout B) -> Transpose (non-blocked layout C) https://github.com/openvinotoolkit/oneDNN/blob/2ead5d4fe5993a797d9a7a4b8b5557b96f6ec90e/src/cpu/reorder/cpu_reorder_regular_f32_f32.cpp#L82

In the context of the pattern and current implementation of Reorder and Transpose, there is no reverse order for this situation.

Blocked layout B could only be ncsp8c or ncsp16c, and non-blocked layout C is ncsp; layout A could only be nspc.

2. But the second situation is still possiblely asked for the reverse order. Please refer to src/plugins/intel_cpu/tests/unit/graph/merge_transpose_reorder_test.cpp.

Reply:
LGTM